pkg/controller/core.oam.dev/v1alpha2/application/assemble/assemble.go

```diff
@@ -0,0 +1,281 @@
+	}
+	return nil
+}
+
+func (o *Options) setWorkloadName(wl *unstructured.Unstructured, compName string) {
+	// use component name as workload name
+	// override the name set in render phase if exist
+	wl.SetName(compName)
+}
+
+func (o *Options) setTraitName(trait *unstructured.Unstructured, compName, traitType string, compTrait *v1alpha2.ComponentTrait) {
+	// NOTE Comparing to AppConfig, Assemble can not use existing name recorded in AppConifg's status
+	// only set generated name when name is unspecified
+	// it's by design to set arbitrary name in render phase
+	if len(trait.GetName()) == 0 {
+		traitName := util.GenTraitName(compName, compTrait, traitType)
+		trait.SetName(traitName)
+	}
+}
+
+// workload and trait in the same component both will have these lables
+func (o *Options) generateCommonLables(compName, compRevisionName string) map[string]string {
+	lables := map[string]string{
+		oam.LabelAppName:              o.AppName,
+		oam.LabelAppRevision:          o.AppRevision.Name,
+		oam.LabelAppRevisionHash:      o.AppRevision.Labels[oam.LabelAppRevisionHash],
+		oam.LabelAppComponent:         compName,
+		oam.LabelAppComponentRevision: compRevisionName,
+	}
+	// pass application's all labels to workload/trait
+	return util.MergeMapOverrideWithDst(lables, o.AppLables)
+}
+
+func (o *Options) setWorkloadLables(wl *unstructured.Unstructured, commonLables map[string]string) {
+	// add more workload-specific labels here
+	util.AddLabels(wl, map[string]string{oam.LabelOAMResourceType: oam.ResourceTypeWorkload})
+	util.AddLabels(wl, commonLables)
+
+	/* NOTE a workload has these possible labels
+	   app.oam.dev/app-revision-hash: ce053923e2fb403f
+	   app.oam.dev/appRevision: myapp-v2
+	   app.oam.dev/component: mycomp
+	   app.oam.dev/name: myapp
+	   app.oam.dev/resourceType: WORKLOAD
+	   app.oam.dev/revision: mycomp-v2
+	   workload.oam.dev/type: kube-worker
+	*/
+}
+
+func (o *Options) setTraitLables(trait *unstructured.Unstructured, commonLables map[string]string) {
+	// add more trait-specific labels here
+	util.AddLabels(trait, map[string]string{oam.LabelOAMResourceType: oam.ResourceTypeTrait})
+	util.AddLabels(trait, commonLables)
+
+	/* NOTE a trait has these possible labels
+	   app.oam.dev/app-revision-hash: ce053923e2fb403f
+	   app.oam.dev/appRevision: myapp-v2
+	   app.oam.dev/component: mycomp
+	   app.oam.dev/name: myapp
+	   app.oam.dev/resourceType: TRAIT
+	   app.oam.dev/revision: mycomp-v2
+	   trait.oam.dev/resource: service
+	   trait.oam.dev/type: ingress // already added in render phase
+	*/
+}
+
+// workload and trait both have these annotations
+func (o *Options) setAnnotations(obj *unstructured.Unstructured) {
+	// pass application's all annotations
+	util.AddAnnotations(obj, o.AppAnnotations)
+	// remove useless annotations for workload/trait
+	util.RemoveAnnotations(obj, []string{
+		oam.AnnotationAppRollout,
+		oam.AnnotationRollingComponent,
+		oam.AnnotationInplaceUpgrade,
+	})
```

Comment:
Does the logic truly needed? These annotations are pass from AC set by appRollout. After refactor AppRollout won't set these label.

Reply:
It works the same as how AppConfig works now. These annotations are just removed from workloads/traits, rollout controller will not consume them from workloads/traits directly, right?

Reply:
Yes,AppRollout control appContext templating workload by setting these annotation.After remove appContext rollout won't set them directly to workload/trait anymore.